agent/structs/config_entry_gateways.go

```diff
@@ -0,0 +1,124 @@
+package structs
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/acl"
+)
+
+// IngressGatewayConfigEntry manages the configuration for an ingress service
+// with the given name.
+type IngressGatewayConfigEntry struct {
+	Kind string
+	Name string
+
+	Listeners []IngressListener
+
+	EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+type IngressListener struct {
+	Port     int
+	Protocol string
+
+	Services []IngressService
+}
+
+type IngressService struct {
+	Name          string
+	ServiceSubset string
+
+	EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+}
+
+func (e *IngressGatewayConfigEntry) GetKind() string {
+	return IngressGateway
+}
+
+func (e *IngressGatewayConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+
+	return e.Name
+}
+
+func (e *IngressGatewayConfigEntry) Normalize() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	e.Kind = IngressGateway
+	for _, listener := range e.Listeners {
+		listener.Protocol = strings.ToLower(listener.Protocol)
+	}
+
+	e.EnterpriseMeta.Normalize()
+
+	return nil
+}
+
+func (e *IngressGatewayConfigEntry) Validate() error {
+	declaredPorts := make(map[int]bool)
+	for _, listener := range e.Listeners {
+		if _, ok := declaredPorts[listener.Port]; ok {
+			return fmt.Errorf("port %d declared on two listeners", listener.Port)
+		}
+		declaredPorts[listener.Port] = true
+
+		for _, s := range listener.Services {
+			if s.Name == "*" && listener.Protocol != "http" {
+				return fmt.Errorf("Wildcard service name is only valid for protocol = 'http' (listener on port %d)", listener.Port)
+			}
+			if s.NamespaceOrDefault() == WildcardSpecifier {
```

Comment:
Namespace validation should happen in consul-enterprise instead. 

OSS users should see an error if an error like this one:
https://github.com/hashicorp/consul/blob/master/agent/http_oss.go#L15

Reply:
Any thoughts on what kind of function we could use here? Maybe something like this:

```go
func (m *EnterpriseMeta) NamespaceNoWildcard() (string, error) {
	return "", fmt.Errorf("Namespaces is a Consul Enterprise feature")
}
```

and call with:
```go
if _, err := s.NamespaceNoWildcard() {
	return err
}
```

We do something like this here:
https://github.com/hashicorp/consul/blob/master/agent/http_oss.go#L23

Reply:
Can you explain more why you think this validation should go in enterprise?

We are only validation the wildcard specifier, and technically it is impossible for this to be anything but "default" in OSS.

Reply:
Oh, I see. I thought it would be possible to input a non-default namespace, but the `EnterpriseMeta` in `IngressService` will prevent that.

Nevermind then, this seems OK.